Introduce S3 client API

tgrabiec opened the pull request:
Example usage:

```
auto client = s3::make_client(
      s3::make_basic_connection_factory("127.0.0.1", 9000));
co_await client->create_file("/mybucket/test_file");
co_await client->open("/mybucket/test_file");
```

tgrabiec committed:
Introduce S3 client API

psarna commented on s3/s3.cc:
```diff
@@ -0,0 +1,347 @@
+    log.debug("conn {}: {} {} {}", fmt::ptr(&*con), method, path, headers);
+
+    sstring method_ = method;
+    sstring path_ = path;
```
Looks like `method` and `path` can simply be taken by value in the function header if we need copies anyway

tgrabiec replied:
It's an implementation detail, we could later optimize this to not copy.
It's also a bit easier to pass references through layers, cause I don't need to care about std::move() and mutable

psarna commented on s3/s3.cc:
```diff
@@ -0,0 +1,347 @@
+
+    auto& in = con->in();
+    line_consumer lc;
+    co_await in.consume(lc);
```
Perhaps we can just use Seastar's http response parser to make sure that we don't miss any edge cases? Or do you expect performance penalty? Example: https://github.com/avikivity/scylla/blob/9caf85f64afec660fbac137286c8b610e86e5432/locator/gce_snitch.cc#L126-L128

It's quite neat, since it also implements a `Consumer` trait and gives you a parsed response right away.

tgrabiec replied:
I forgot that we have it, will give it a try

psarna replied:
Now that I look at it, I think we never store the response code anywhere during parsing, just the HTTP version and the headers... so it's semi-useful. It would be extremely easy to add the capability of storing the response code when parsing, but I don't think it's done now.

psarna replied:
I'll take a look if I can quickly add it, I'll need it for websockets anyway

psarna commented on s3/s3.cc:
```diff
@@ -0,0 +1,347 @@
+
+seastar::logger log("s3");
+
+struct http_response {
```
Seastar already defines a http_response struct and a parser (more info in the comment below). Perhaps we don't need to reinvent the wheel and we can just reuse it.

tgrabiec replied:
I figured that the seastar one is too primitive for my purpose. Perhaps we could compose it. But, given that it's a hackaton, I wouldn't waste time on it now.

avikivity commented on s3/s3.cc:
```diff
@@ -0,0 +1,347 @@
+    virtual future<> truncate(uint64_t length) override { unsupported(__FUNCTION__); }
+    virtual future<> discard(uint64_t offset, uint64_t length) override { unsupported(__FUNCTION__); }
+    virtual future<> allocate(uint64_t position, uint64_t length) override { unsupported(__FUNCTION__); }
+    virtual subscription<directory_entry> list_directory(std::function<future<>(directory_entry)>) override { unsupported(__FUNCTION__); }
```
We need all those! Or the node won't bootstrap.

tgrabiec replied:
Isn't list_directory() going to be replaced with a read from system tables where we store the list of sstable object names?

tgrabiec replied:
The other ones cannot be implemented, so the `storage` abstraction should make sure these are not called for this storage type.

avikivity replied:
> Isn't list_directory() going to be replaced with a read from system tables where we store the list of sstable object names?

Yes. I thought to keep this unchanged, but it's probably impossible since there's no other way to tell which sstables are ours.


@ManManson please add an interface to query all sstables that are co-owned by a node.

tgrabiec committed:
utils: memory_data_sink: Override mandatory buffer_size()

The default implementation aborts. The class has bit rot because it was unused.
